### PR TITLE
Issues #23, #16

### PR DESCRIPTION
**Pull request by tdorianh**
This addresses issues #23, #16

### CODE REVIEW

File: constellation_ticketbox/forms.py [hunk 1]
```diff
@@ -23,9 +26,11 @@ def __init__(self, *args, **kwargs):
 
     class Meta:
         model = Ticket
-        fields = ['title', 'body', 'anonymous', 'status',]
+        fields = ['title', 'body', 'anonymous', 'status', ]
```
Comment (the-maldridge) [inline constellation_ticketbox/forms.py]: If there is no next element, the list should not have a continuation comma.

File: constellation_ticketbox/models.py [hunk 2]
```diff
@@ -4,16 +4,18 @@
...
     ticket = models.ForeignKey('Ticket', blank=True, null=True)
     owner = models.ForeignKey(User)
     author = models.CharField(max_length=128, default='Anonymous')
     body = models.TextField(blank=False, null=False)
     timestamp = models.DateTimeField(auto_now_add=True)
-    
+
+
 class Ticket(models.Model):
     STATUS_CHOICES = (
```
Comment (the-maldridge) [inline constellation_ticketbox/models.py]: Why is this an enum?  Why can't these just be strings?

File: constellation_ticketbox/tests.py [hunk 3]
```diff
@@ -1,3 +1,2 @@
-from django.test import TestCase
```
Comment (the-maldridge) [inline constellation_ticketbox/tests.py:1]: If you have no tests this file should be removed.

File: constellation_ticketbox/urls.py [hunk 4]
```diff
@@ -4,63 +4,63 @@
 
 urlpatterns = [
 
-# =============================================================================
-# View Routes
-# =============================================================================
+    # =============================================================================
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: This should be left aligned to match the rest of the constellation suite.

File: constellation_ticketbox/urls.py [hunk 5]
```diff
@@ -4,63 +4,63 @@
...
 
-    url(r'^view/ticket/(?P<ticket_id>\d+)$', 
+    url(r'^view/ticket/(?P<ticket_id>\d+)$',
         views.view_ticket,
         name="view_ticket"),
 
-# =============================================================================
-# Management Routes
-# =============================================================================
+    # =============================================================================
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left align

File: constellation_ticketbox/urls.py [hunk 6]
```diff
@@ -4,63 +4,63 @@
...
 
-    url(r'^manage/box/(?P<box_id>\d+)/edit$', 
+    url(r'^manage/box/(?P<box_id>\d+)/edit$',
         views.manage_box_edit,
         name="manage_box_edit"),
 
-# =============================================================================
-# API Routes for the v1 API
-# =============================================================================
+    # =============================================================================
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left Align

File: constellation_ticketbox/urls.py [hunk 7]
```diff
@@ -4,63 +4,63 @@
...
+    # API Routes for the v1 API
+    # =============================================================================
 
-# -----------------------------------------------------------------------------
-# API Routes related to Box Operations
-# -----------------------------------------------------------------------------
+    # -----------------------------------------------------------------------------
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left align

File: constellation_ticketbox/urls.py [hunk 8]
```diff
@@ -72,38 +72,38 @@
         views.api_v1_box_closed_tickets,
         name="api_v1_box_closed_tickets"),
 
-# -----------------------------------------------------------------------------
-# API Routes related to Ticket Operations
-# -----------------------------------------------------------------------------
+    # -----------------------------------------------------------------------------
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left align

File: constellation_ticketbox/urls.py [hunk 9]
```diff
@@ -72,38 +72,38 @@
...
 
     url(r'^api/v1/ticket/(?P<ticket_id>\d+)/update-status$',
         views.api_v1_ticket_update_status,
         name="api_v1_ticket_update_status"),
 
-# -----------------------------------------------------------------------------
-# API Routes related to Reply Operations
-# -----------------------------------------------------------------------------
-    
+    # -----------------------------------------------------------------------------
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left align

File: constellation_ticketbox/urls.py [hunk 10]
```diff
@@ -72,38 +72,38 @@
...
+
     url(r'^api/v1/ticket/(?P<ticket_id>\d+)/reply/create$',
         views.api_v1_reply_create,
         name="api_v1_reply_create"),
 
 
-# -----------------------------------------------------------------------------
-# Dashboard routes
-# -----------------------------------------------------------------------------
+    # -----------------------------------------------------------------------------
```
Comment (the-maldridge) [inline constellation_ticketbox/urls.py]: Left align

File: constellation_ticketbox/views.py [hunk 11]
```diff
@@ -241,7 +246,7 @@ def api_v1_box_unarchive(request, box_id):
 def api_v1_box_open_tickets(request, box_id):
     '''Retrieve all unarchived tickets for box'''
     ticketObjects = Ticket.objects.filter(box=Box.objects.get(pk=box_id),
-                    archived=False)
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: I am having trouble believing that this is flake8 compliant indentation.

File: constellation_ticketbox/views.py [hunk 12]
```diff
@@ -255,7 +260,7 @@ def api_v1_box_open_tickets(request, box_id):
 def api_v1_box_closed_tickets(request, box_id):
     '''Retrieve all archived tickets for box'''
     ticketObjects = Ticket.objects.filter(box=Box.objects.get(pk=box_id),
-                    archived=True)
+                                          archived=True)
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: indentation...

File: constellation_ticketbox/views.py [hunk 13]
```diff
@@ -279,19 +285,19 @@ def api_v1_ticket_create(request, box_id):
         newTicket.box = get_object_or_404(Box, pk=box_id)
         newTicket.owner = request.user
         newTicket.anonymous = ticketForm.cleaned_data['anonymous']
-        if (newTicket.anonymous == False):
+        if (newTicket.anonymous is False):
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: an explicit check for False here is not needed.  Check for not newTicket.anonymous instead.

File: constellation_ticketbox/views.py [hunk 14]
```diff
@@ -301,14 +307,16 @@ def api_v1_ticket_replies(request, ticket_id):
...
+        replyObjects = Reply.objects.filter(
+            ticket=Ticket.objects.get(pk=ticket_id))
         if replyObjects:
             replies = serializers.serialize('json', replyObjects)
             return HttpResponse(replies)
         else:
             return HttpResponseNotFound("There are no replies to this ticket.")
     else:
-        return HttpResponseNotFound("You do not have permissions to see this ticket.")
+        return HttpResponseNotFound("You do not have permissions " +
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: These lines will be joined implicitly, there is no need for the '+'

File: constellation_ticketbox/views.py [hunk 15]
```diff
@@ -320,17 +328,18 @@ def api_v1_ticket_update_status(request, ticket_id):
         box = ticket.box
         box_perms = get_perms(request.user, box)
 
-        if ((ticket.owner == request.user) or ('action_manage_tickets' in box_perms)):
+        if ((ticket.owner == request.user) or
+                ('action_manage_tickets' in box_perms)):
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: over indented.

File: constellation_ticketbox/views.py [hunk 16]
```diff
@@ -320,17 +328,18 @@ def api_v1_ticket_update_status(request, ticket_id):
...
                 newStatus = ticketForm.cleaned_data['status']
                 # prevent duplicate status changes
                 if (newStatus == ticket.status):
-                    return redirect(reverse("view_ticket", args=[ticket_id,]))
+                    return redirect(reverse("view_ticket", args=[ticket_id, ]))
 
                 # prevent owners from re-opening their own tickets
-                if (ticket.owner == request.user and ticket.archived == True):
-                    return redirect(reverse("view_ticket", args=[ticket_id,]))
+                if (ticket.owner is request.user and ticket.archived is True):
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: ticket.archived is already True on its own, this comparison is unnecessary.

File: constellation_ticketbox/views.py [hunk 17]
```diff
@@ -343,22 +352,23 @@ def api_v1_ticket_update_status(request, ticket_id):
                 newReply.ticket = ticket
                 newReply.owner = request.user
                 # preserve anonymity
-                if (ticket.anonymous == True and ticket.owner == request.user):
+                if (ticket.anonymous is True and ticket.owner is request.user):
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Unnecessary comparison to literal.

File: constellation_ticketbox/views.py [hunk 18]
```diff
@@ -379,22 +389,25 @@ def api_v1_reply_create(request, ticket_id):
             newReply.owner = request.user
 
             # preserve anonymity
-            if (ticket.anonymous == True and request.user == ticket.owner):
+            if (ticket.anonymous is True and request.user is ticket.owner):
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Unnecessary literal comparison

File: constellation_ticketbox/views.py [hunk 19]
```diff
@@ -379,22 +389,25 @@ def api_v1_reply_create(request, ticket_id):
...
 
             newReply.ticket = get_object_or_404(Ticket, pk=ticket_id)
             try:
                 newReply.save()
-                return HttpResponse(serializers.serialize('json', [newReply,]))
+                return HttpResponse(serializers.serialize('json',
+                                                          [newReply, ]))
             except:
-                return HttpResponseServerError("Could not save reply at this time")
+                return HttpResponseServerError("Could not save reply" +
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Unnecessary explicit line join

File: constellation_ticketbox/views.py [hunk 20]
```diff
@@ -379,22 +389,25 @@ def api_v1_reply_create(request, ticket_id):
...
+                                               " at this time")
         else:
             return HttpResponseBadRequest("Invalid Form Data!")
     else:
-        return HttpResponseBadRequest("You don't have permissions to reply to this ticket.")
-    
+        return HttpResponseBadRequest("You don't have permissions " +
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Unnecessary explicit line join

File: constellation_ticketbox/views.py [hunk 21]
```diff
@@ -343,22 +352,23 @@ def api_v1_ticket_update_status(request, ticket_id):
...
+                newReply.body += newStatus + '\'.'
                 newReply.save()
-                return redirect(reverse("view_ticket", args=[ticket_id,]))
+                return redirect(reverse("view_ticket", args=[ticket_id, ]))
             except AttributeError:
                 return HttpResponseServerError("Invalid Ticket ID!")
-        else: 
-            return HttpResponseBadRequest("You don't have permissions to edit this ticket's status.")
+        else:
+            return HttpResponseBadRequest("You don't have permissions " +
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Unnecessary explicit line join

File: constellation_ticketbox/views.py [hunk 22]
```diff
@@ -320,17 +328,18 @@ def api_v1_ticket_update_status(request, ticket_id):
         box = ticket.box
         box_perms = get_perms(request.user, box)
 
-        if ((ticket.owner == request.user) or ('action_manage_tickets' in box_perms)):
+        if ((ticket.owner == request.user) or
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py:335]: Comparison of objects using `==` instead of `is`

File: constellation_ticketbox/views.py [hunk 23]
```diff
@@ -92,21 +93,23 @@ def view_ticket(request, ticket_id):
...
+            'form': replyForm,
+            'status_form': ticketForm,
             'id': ticket_id,
             'template_settings': template_settings,
             'ticket': ticket,
             'box': box,
         })
     else:
-        return HttpResponseNotFound("You do not have permissions for this ticket.")
+        return HttpResponseNotFound("You do not have permissions " +
```
Comment (the-maldridge) [inline constellation_ticketbox/views.py]: Lines are already implicitly joined.

File: constellation_ticketbox/forms.py [hunk 24]
```diff
@@ -2,7 +2,9 @@
 from .models import Ticket
 from django.forms import ModelForm
 
+
 class BoxForm(ModelForm):
+
```
Comment (the-maldridge) [inline constellation_ticketbox/forms.py]: There should be a docstring here

File: constellation_ticketbox/forms.py [hunk 25]
```diff
@@ -15,6 +17,7 @@ class Meta:
 
 
 class TicketForm(ModelForm):
+
```
Comment (the-maldridge) [inline constellation_ticketbox/forms.py]: There should be a docstring here

**Review by the-maldridge — CHANGES_REQUESTED** [review]
Lots of places are  casting to lists with commas that are not always necessary.  If you do not have another item in the list, there should be no comma and no space after the list element.

File: constellation_ticketbox/views.py (hunk 22 above)
Comment (tdorianh) [reply constellation_ticketbox/views.py:335]: For some reason using `is` will cause this test to always fail, but it has the correct functionality using `==`.

File: constellation_ticketbox/models.py (hunk 2 above)
Comment (tdorianh) [reply constellation_ticketbox/models.py]: Status is currently used to determine whether or not a ticket should be archived upon status update. 

Should this be changed to require an explicit archiving when a ticket status is updated, and allow users to enter anything into the status field?

**Comment by tdorianh** [comment]
Commit 04dcd04 removes the previous method of archiving tickets by selecting statuses from a limited list. In this version, archiving and status updates are separate buttons, allowing managing groups to update a status without changing the open/closed archive state of the ticket (such as for status messages like "In progress"). Status is also now simply strings. A button on the ticket page has been added for archiving and unarchiving.

**Pull request merged**